### ISSUE TITLE
Auto detect compiler version

### ISSUE THREAD

**Issue by MikeMitterer** [conan-io/conan#762]
Say I have clang 8.0 or gcc 5.4 installed on my Mac, the package I want is not available for these compilers. I have to go to the according package to figure out the supported compiler and the I have to specify it on the command line - a bit annoying...

**Comment by memsharded** [comment]
Not sure what you mean. The compiler version is defined by your installed compiler. Packages usually do not restrict compiler or compiler versions, though the recipes might fail to build from sources if there is something that doesn't build or is not prepared for that specific compiler. Usually not necessary to dig into the package recipes, just provide the ``--build`` option. Could you please explain a bit further? Thanks!

**Comment by MikeMitterer** [comment]
Take this sample: http://blog.conan.io/2016/05/10/Programming-C++-with-the-4-Cs-Clang,-CMake,-CLion-and-Conan.html

"Set up the project"
```bash
conan install .. -s compiler=clang -s compiler.version=3.6 -s compiler.libcxx=libstdc++ --build=missing
```
I don't get it why I have to specify the compiler, the compiler.version asf. Especially if it fails if I specify my compiler version which is 8.0 (clang on El Caption) For this concrete sample I had to set the compiler.version to 7.0 - VERY strange!   

In my opinion conan should first check if the env-vars CC and CXX are set.
If not - it should check the version for c++, g++ or clang - after these checks it knows the OS and the compiler-version I use.
compiler.libcxx=libstdc++ could be the default option for clang
build=missing should also be the default

Now conan knows everything it needs to collect the necessary libs.
If my compiler is clang 8.0 on OSX it should find e.g. Poco apple-clang 7.3 which is the latest available Poco-Version available 

To summarize what I want:
conanfile.txt
```
[requires]
Poco/1.7.2@lasote/stable

[generators]
cmake
```

`conan install .. `

Thanks!

OT: I think the way you sort your packages under "Filter" is a bit confusion
compiler-version from latest to the oldest...
https://conan.io/source/Poco/1.7.3/lasote/stable - go to Macos. You see 6.1, 7.3, 7.0... 7.3 Release - looks strange. I think the most common scenario is that I look for the latest available version

Your Tabs (Macos, Linux and Windows) are also a bit confusing - this assumes that just because I'm on Mac that I use clang - why not gcc. On Linux - why not clang?

**Comment by memsharded** [comment]
Hi Mike,

About the order of packages under "Filter", some tips:
- Under different OSs, it doesn't assume any compiler, it just lists the compiler used to generate those packages. If there are packages in Linux compiled with clang, they will be listed there.
- It is not true that people will look in general for the latest version. Actually, in systems like linux, gcc 4.9 is a very mainstream compiler version.
- The order is quite irrelevant for me, now that a package query is available. If I want to check for my compiler version I write "compiler.version=4.9" and done. Much more convenient than scrolling down to look for it.

Please try it, if you feel that listing the package binaries in order is still very necessary, please open a specific issue for it.

**Comment by memsharded** [comment]
About the install issue, I still don't fully get it. Please let me explain a few things to see if it helps:

- The ``conan install`` is and should be decoupled from any specific setup of the machine. It is true, that for building from sources, it might honor CC and CXX variables (which should be better specified in profiles, rather than globally). You can actually install as many packages as you want, even if you dont have the compilers installed. That is a typical procedure for remote management/sync. In a bare CI server, you might be able to install all binaries from multiple compilers, versions (and even OS), then upload them to another remote.

- Auto-detection of machine dev setup is run at first conan run, and stored as defaults in ``.conan/conan.conf``. Of course you can modify there your defaults. These defaults avoid having to type long command lines in most cases.

- The ``profiles`` feature is also very handy to summarize configurations: http://docs.conan.io/en/latest/reference/config_files.html#profiles

- ``build=missing`` should not be the default. This was decided (always a trade-off) after numerous feedback from users. You are installing a package you thought that you had already built in CI, but for some reason something fail. Users don't want that a build from source will automatically start in their machines or in CI without  noticing. So the decision balanced towards building from source be **opt-in** not **opt-out**. In perspective, we still think it was the best decision (the best for most use cases, not the best for all cases for all developers)

- The ABI compatibility is a big issue. It cannot be assumed that a package compiled in previous versions will be ABI compatible with modern ones. May be in C, but rarely in C++. So the default is that if you are compiling with apple-clang 8.0, your dependencies will have to be compiled with apple-clang 8.0, or if they are not existing, they will have to be built from sources. More importantly if they are majors! It might be possible that gcc 6.1 packages can be linked from gcc 6.2, but jumping in major version is typically a no. There are ways to specify ABI compatibility, but that is a responsibility of the package creator. Using the ``conan_info()`` method, package creators can specify the compatibility of their binaries. For example, if the Poco package creator specify something like ``self.info.settings.compiler.version = "7.X"``, so if you were using lets say "7.5", you will be able to link against a package built with 7.3. But this is always the responsibility of the package creator.

- If you are sure and want to override the settings for a package, you can do it, settings per package can be also specified: http://docs.conan.io/en/latest/reference/commands.html#settings

Please check if this helps or clarifies things. 
Another tip: it is much better if you keep github issues very focused and specific. It really helps to address them faster, are read by more interested users, and are easier to track progress, conversation and decisions. No problem to open several of them.

Thanks for your feedback!

**Comment by ovidiub13** [comment]
@MikeMitterer can we close this issue? Has your question been clarified?

**Comment by MikeMitterer** [comment]
Yes - please + Thanks!
